### PR TITLE
publish grammar without left-recursion; add tooling to make it so

### PR DESCRIPTION
**Pull request by dneto0**

### CODE REVIEW

**Comment by dneto0** [comment]
For #3286 but unfinished.

File: wgsl/wgsl.recursive.bs.include [hunk 1]
```diff
@@ -0,0 +1,693 @@
...
+<div class='syntax' noexport='true'>
+  <dfn for='right_syntax'>additive_operator</dfn>:
+
+ | [=syntax/minus=]
+
+ | [=syntax/plus=]
+</div>
+
+<div class='syntax' noexport='true'>
+  <dfn for='right_syntax'>address_space</dfn>:
```
Comment (dj2) [inline wgsl/wgsl.recursive.bs.include:20]: What does `right_syntax` mean?

File: wgsl/wgsl.recursive.bs.include [hunk 2]
```diff
@@ -0,0 +1,693 @@
...
+</div>
+
+<div class='syntax' noexport='true'>
+  <dfn for='right_syntax'>member_ident</dfn>:[=syntax/ident_pattern_token=]
+</div>
+
+<div class='syntax' noexport='true'>
+  <dfn for='right_syntax'>multiplicative_expression</dfn>:
+
+ | [=right_syntax/multiplicative_expression=]
```
Comment (dj2) [inline wgsl/wgsl.recursive.bs.include:370]: Something has gone wrong here ....

File: wgsl/wgsl.recursive.bs.include [hunk 3]
```diff
@@ -0,0 +1,693 @@
...
+
+ | [=syntax/modulo=]
+
+ | [=syntax/star=]
+</div>
+
+<div class='syntax' noexport='true'>
+  <dfn for='right_syntax'>optionally_typed_ident</dfn>:
+
+ | [=right_syntax/optionally_typed_ident=]
```
Comment (dj2) [inline wgsl/wgsl.recursive.bs.include:386]: This is also weirdly collapsed

File: wgsl/wgsl.recursive.bs.include (hunk 1 above)
Comment (dneto0) [reply wgsl/wgsl.recursive.bs.include:20]: It's a tag, other than "syntax",  for differentiating kinds of definitions.

e.g. "expression" has a main body definition, a definition "for" syntax (the main grammar), and "right syntax" which is for right-recursive grammar, or recursive descent.  I guess I should change it to recursive descent...

Comment (dneto0) [reply wgsl/wgsl.recursive.bs.include:20]: Screen Shot 2022-08-15 at 2 35 48 PM

Comment (dj2) [reply wgsl/wgsl.recursive.bs.include:20]: But, why is the definition for `right_syntax` "address_space" ?

File: wgsl/wgsl.recursive.bs.include (hunk 2 above)
Comment (dneto0) [reply wgsl/wgsl.recursive.bs.include:372]: yep.  The left-factoring code was rewriting the original rule, then immediately reabsorbing. 
Fixed now.

File: wgsl/wgsl.recursive.bs.include (hunk 1 above)
Comment (dneto0) [reply wgsl/wgsl.recursive.bs.include]: The "for" tag is "in what context is this definition being made".
So it's address_space "for" recursive descent grammar.

Comment (dneto0) [reply wgsl/wgsl.recursive.bs.include]: This is what "address_space" looks like in the index:

Screen Shot 2022-08-15 at 3 15 32 PM

Comment (dj2) [reply wgsl/wgsl.recursive.bs.include]: I see, interesting.

**Review by dj2 — APPROVED** [review]

**Comment by dneto0** [comment]
`undercore equal expression` has disappeared(!)

File: wgsl/analyze/lalr.py [hunk 4]
```diff
@@ -74,13 +90,62 @@ def main():
...
+        # Bring together with other rules starting with attribute
+        g.inline_when_toplevel_prefix({'global_constant_decl'})
+
+        g.inline_single_starrable()
+
+        g.refactor_post('ident')
+        g.rotate_one_or_mores()
+
+        print(g.pretty_str(po))
         sys.exit(0)
```
Comment (dj2) [inline wgsl/analyze/lalr.py:137]: Should this be here?

File: wgsl/wgsl.recursive.bs.include [hunk 5]
```diff
@@ -376,12 +372,6 @@
  | [=syntax/star=]
 </div>
 
-<div class='syntax' noexport='true'>
-  <dfn for='recursive descent syntax'>optionally_typed_ident</dfn>:
```
Comment (dj2) [inline wgsl/wgsl.recursive.bs.include]: I don't think this should have been inlined?

File: wgsl/analyze/lalr.py (hunk 4 above)
Comment (dneto0) [reply wgsl/analyze/lalr.py]: No. Removed now.
thanks

File: wgsl/wgsl.recursive.bs.include (hunk 5 above)
Comment (dneto0) [reply wgsl/wgsl.recursive.bs.include]: Added it to the exclusion list of things to be inlined.  Reads better that way.

**Review by mehmetoguzderin — APPROVED** [review]
Great analysis, LGTM. (I wonder if some parts of output can incorporate into the original form)

**Pull request merged**